### DISCUSSION TITLE
`depends_on :arch` evaluated as "AND" for multiple architectures

### DISCUSSION THREAD

**Discussion by radeksimko** [Homebrew/discussions#1595] [Tap maintenance and Homebrew development]
👋🏻  I appreciate that in many cases casks and formulae are manually maintained and in these cases it's easy to write down the right stanza for certain edge cases, where only certain architecture is supported.

In GoReleaser we automate publishing formulae and we'd like to do the right thing in terms of setting the architecture requirements, to avoid publishing formulae which result in none `url` when evaluated on unsupported architectures.

As a result we [attempted](https://github.com/goreleaser/goreleaser/pull/2261) to generate `depends_on :arch` as an array with _all_ supported architectures. This however lead to us to realize that the following stanza

```ruby
depends_on arch: [:x86_64, :aarch64]
```
results in unsatisfiable requirement and error as reported in https://github.com/goreleaser/goreleaser/issues/2270

I can't see any meaningful use for requiring multiple architectures _at the same time_ - the host would always be identified with a single architecture, so that leads me to believe it's a bug.

Is this a feature or a bug?

**Comment by carlocab** [comment]
I don't think this is a bug. This matches the semantics of the following declaration, for example:
```ruby
depends_on "foo" => [:build, :test]
```
which says that `foo` is both a build and a test dependency, and not that `foo` is at least one of a build- or test-dependency.

You are right in pointing out that it's not useful to require multiple architectures at the same time, but I suspect that if we want the ability to declare an array of arch symbols as specifying the possible compatible architectures, then we'd need new syntax for that.

**Reply by radeksimko** [reply]
I see, so in that case I would expect to see a more upfront error to say that multiple architectures is basically invalid stanza - which would IMO avoid similar potential confusions.

Until a new syntax is considered/introduced, would something like this work? 🤔 

```ruby
depends_on arch: [:x86_64] if Hardware::CPU.intel? && Hardware::CPU.is_64_bit?
depends_on arch: [:aarch64] if Hardware::CPU.arm? && Hardware::CPU.is_64_bit?
```

**Reply by radeksimko** [reply]
Actually we'd probably need the opposite - evaluate `depends_on` if the condition behind it _is not_ satisfied.

**Reply by carlocab** [reply]
Do you have an example of a formula where you might want to declare a dependency on one of a set of possible arches?

**Reply by SMillerDev** [reply]
> Until a new syntax is considered/introduced, would something like this work? 🤔

What you're describing here seems to be roughly how bottles work. Wouldn't it be easier to bottle your software?

**Reply by radeksimko** [reply]
@SMillerDev Is that an option for 3rd party taps? https://github.com/Homebrew/brew/blob/master/docs/Bottles.md#creation suggests it is not?

**Reply by nandahkrishna** [reply]
You might want to look at this: https://brew.sh/2020/11/18/homebrew-tap-with-bottles-uploaded-to-github-releases/#post.

**Reply by carlocab** [reply]
BrewTestBot just runs `brew` commands (e.g. `brew install --build-bottle foo && brew bottle foo`). You can also have your own BrewTestBot: https://brew.sh/2020/11/18/homebrew-tap-with-bottles-uploaded-to-github-releases/

**Edit:** scooped by @nandahkrishna once again...

**Reply by radeksimko** [reply]
> Do you have an example of a formula where you might want to declare a dependency on one of a set of possible arches?

I think it's less about a human-maintained example, but more about ease of generating the formulae.

https://github.com/goreleaser/goreleaser/blob/master/internal/pipe/brew/template.go

I suppose we could avoid generating `depends_on` when **all** architectures are supported, which kind of solves the problem for macOS as we'd be left with just one of two architectures (intel/arm) for `depends_on` if any, but on Linux we have 3 to deal with, so the formula generator would need to account for that anyway somehow.

**Reply by radeksimko** [reply]
Thanks for the links.

It looks useful, but from what I can tell bottles are basically just "pre-compiled binaries", right? If so, then that doesn't sound like something we'd want to be generating from GoReleaser as in that context it's GoReleaser compiling the binaries - typically in more complex way than a bottle bot? e.g. with code signing, custom build flags etc.

It's probably not impossible to set it up that way, but it feels like a lot of effort (basically reimplementing GoReleaser as a bottle bot) considering the original problem which was just a clearer communication of architecture requirements.

**Reply by radeksimko** [reply]
The concept of bottles generally looks useful, but I think we'd probably need to find a way of just leveraging the binaries/archives that we already have + corresponding hashes and just linking these from the formula?

**Reply by radeksimko** [reply]
Ha, I think this is what I was looking for:

https://github.com/Homebrew/brew/blob/master/docs/Bottles.md#root-url-root_url

**Reply by carlocab** [reply]
> I think it's less about a human-maintained example, but more about ease of generating the formulae.

I'm aware of that, but showing me your template isn't really useful, as it doesn't really give me a clearer idea of the problem you're trying to solve.

**Reply by radeksimko** [reply]
@carlocab here is an example - assuming that we wanted to drop support for arm 32bit on Linux, then we'd end up with the following:

```ruby
class Goreleaser < Formula
  desc "Deliver Go binaries as fast and easily as possible"
  homepage "https://goreleaser.com"
  version "0.166.2"
  license "MIT"
  bottle :unneeded

  on_macos do
    if Hardware::CPU.intel?
      url "https://github.com/goreleaser/goreleaser/releases/download/v0.166.2/goreleaser_Darwin_x86_64.tar.gz"
      sha256 "ae0917d7642b4cdb21e703fc7019a77678973985728e6efddfa467d1d7b7e168"
    end
    if Hardware::CPU.arm?
      url "https://github.com/goreleaser/goreleaser/releases/download/v0.166.2/goreleaser_Darwin_arm64.tar.gz"
      sha256 "5b4f52dbf9f2355e98abbf1510cf3f66711b5146e38fd0e4d3188b57e1cede12"
    end
  end

  on_linux do
    if Hardware::CPU.intel?
      url "https://github.com/goreleaser/goreleaser/releases/download/v0.166.2/goreleaser_Linux_x86_64.tar.gz"
      sha256 "04aae02a277c48be8ee429e73ca8e1a8ea46cf03cc966fc354ec9cc2e52a8f37"
    end
    if Hardware::CPU.arm? && Hardware::CPU.is_64_bit?
      url "https://github.com/goreleaser/goreleaser/releases/download/v0.166.2/goreleaser_Linux_arm64.tar.gz"
      sha256 "45233abdb6303d2e30c5f7b4a2e7e7f01f35ab3f20f1b4dce9f0a78bbec30daf"
    end
  end

  depends_on "go" => :optional
  depends_on "git"

  def install
    bin.install "goreleaser"
    bash_completion.install "completions/goreleaser.bash" => "goreleaser"
    zsh_completion.install "completions/goreleaser.zsh" => "_goreleaser"
    fish_completion.install "completions/goreleaser.fish"
  end

  test do
    system "#{bin}/goreleaser -v"
  end
end
```

which would presumably result in an error during evaluation on Linux 32bit without any appropriate `depends_on`.

**Reply by carlocab** [reply]
I see. 

One thing you may consider doing here would be to create a custom `Requirement`. This will allow you to define, say a `SupportedArchRequirement` that checks whether the system architecture is one you support. Having done that, the formula can do
```ruby
depends_on SupportedArchRequirement
```
We no longer do `Requirement`s in Homebrew/core, but you can have a look at this tap for an example: https://github.com/gromgit/homebrew-fuse

**Reply by radeksimko** [reply]
I see, that's a useful tip for a workaround to consider.

Regarding the bottling option - the only way that I can imagine us realistically implementing it would be by putting existing binaries "in the bottles", so unless we introduce a 3rd party dependency on another tool it means that GoReleaser would need to produce the bottle archive of a particular structure & content. Is the structure of the bottle archive considered a "stable API" and is it documented somewhere? i.e. is this even advisable avenue to explore or would it be too fragile?

--- 

The more I think about it the more I feel that it would be just easier to introduce something like this to `Homebrew/core`

```ruby
depends_on any_of_arch: [:x86_64, :aarch64]
```
or `any_arch`... I'm not too concerned about the name, but would you accept a PR which adds something like this to `Homebrew/core`?

**Reply by carlocab** [reply]
We'll review a PR for this, but it should entail minimal code changes. A PR that changes
```ruby
depends_on arch: [:x86_64, :aarch64]
```
to mean what you expected it to might be fine as well (even if it has the opposite semantics to other types of usage).

**Reply by caarlos0** [reply]
I tried the following:

```rb
# test.rb

class SupportedArchRequirement < Requirement
  fatal true

  satisfy(build_env: false) { self.class.supported_arch? }

  def message
    "Your platform #{self.class.plat} is not supported by this tap. Supported arches are #{self.class.supported_arches}."
  end

  LINUX_SUPPORTED_ARCHES = [:x86_64, :arm, :aarch64]
  MACOS_SUPPORTED_ARCHES = [:aarch64]

  def self.supported_arch?
    self.supported_arches.include? self.plat
  end

  def self.supported_arches
    return LINUX_SUPPORTED_ARCHES if OS.linux?
    return MACOS_SUPPORTED_ARCHES if OS.mac?
    []
  end

  def self.plat
    case RUBY_PLATFORM
    when /x86_64/ then :x86_64
    when /arm/ then :arm
    when /aarch64/ then :aarch64
    else :dunno
    end
  end
end


class Test < Formula
  desc "Test etc"
  homepage "https://example.com"
  version "1.0.0"
  sha256 "aaaaaa"

  depends_on SupportedArchRequirement

  on_macos do
    if Hardware::CPU.arm?
      url "https://fake.url/macos/arm"
    end
  end

  on_linux do
    url "https://fake.url/linux/all"
  end

  def install
    bin.install "test"
  end

  def caveats
    "Test formula, dont use it. It doesnt do anything"
  end
end
```

```sh
brew install ./test.rb
```

On an intel mac, got:

```
Error: formulae require at least a URL
```

So, seems like it evaluates that before the requirements?

Anyway, adding a "fake" url and trying again:

```sh
Error: Failed to load cask: ./test.rb
Cask 'test' is unreadable: wrong constant name #<Class:0x00007f8b932007f8>
Warning: Treating ./test.rb as a formula.
test: Your platform x86_64 is not supported by this tap. Supported arches are [:aarch64].
Error: An unsatisfied requirement failed this build.
```

I guess the constant error is about my `Requirement`? What's the proper way of fixing it?

**Reply by caarlos0** [reply]
Hmm, seems like putting that in an actual tap fixes the class issue:

```
❯ brew install caarlos0/tap/test
==> Installing test from caarlos0/tap
test: Your platform x86_64 is not supported by this tap. Supported arches are [:aarch64].
Error: An unsatisfied requirement failed this build.
```

**Reply by caarlos0** [reply]
Is that a good way of solving this? 

FWIW, opened a PR adding it to goreleaser: https://github.com/goreleaser/goreleaser/pull/2278

**Comment by ryancurrah** [comment]
@carlocab our company is starting send out M1 macs to developers and this issue has started affecting us. Has there been any progress to support @caarlos0 use case?

**Reply by SMillerDev** [reply]
Well, if you strictly need this feature you can make a pull request. But the conversation above suggests that this isn't really a case that is needed if you use your formulae the same as Homebrew uses its formula.

**Reply by caarlos0** [reply]
I did ask a question [here](https://github.com/Homebrew/discussions/discussions/1595#discussioncomment-817322), but got no reply... so I'm not sure if that's a good way of fixing it...

**Reply by SMillerDev** [reply]
In your personal tap you can do whatever you want so the opinion of homebrew maintainers isn't that relevant here.

**Reply by caarlos0** [reply]
The thing is: GoReleaser generates those taps, and a lot of people use it... so "doing whatever I want" might have bigger implications... That's why I asked that, to see if that's a good way of solving it, of if there's a better way...

**Reply by carlocab** [reply]
There was a PR to try to fix this in https://github.com/Homebrew/brew/pull/11568 but it wasn't finished. Maybe you'd like to pick it up?

Alternatively, maybe a simpler way around this would be to have a single arch-independent tarball (which contains binaries for every arch/os). You can then select the right binary to install inside the `install` method. You might not even need a complicated `if` block to do this if you use helper methods like `Hardware::CPU.arch` and `OS.kernel_name`.

**Reply by carlocab** [reply]
Or I guess another alternative would be to teach GoReleaser to generate bottle artefacts instead, since `brew` knows how to map the host os/arch into the right artefact to use.

**Reply by caarlos0** [reply]
hmm, maybe its worth looking into the bottle thing... any reference recommendations?

**Reply by carlocab** [reply]
Not really, unfortunately. One downside to doing this would be that you would technically be relying on private `brew` internals -- but changing these in a backwards-incompatible way would also break a lot of things (e.g. installs from existing bottles), so it's relatively unlikely to change.

Here's an example of a bottle block you want to generate where the bottles are hosted on a GitHub release:
https://github.com/carlocab/homebrew-personal/blob/062c9226111e2e5fd9ecb8b6c50a30293e07e5dc/Formula/flang.rb#L10-L16

You'll probably want to specify `:any_skip_relocation` in your `bottle` blocks to specify that `brew` doesn't need to do any relocation of your bottles.

And here is the corresponding release that bottles are downloaded from:
https://github.com/carlocab/homebrew-personal/releases/tag/flang-12.0.0_1

Bottles are just tarballs with the extension `bottle.tar.gz`, except when the bottle block specifies a non-zero `rebuild`, in which case the extension is `bottle.$rebuild.tar.gz`. Bottles are designed to be unpacked directly into `HOMEBREW_CELLAR`, so they should already have the necessary directory structure (`bin`, `share`, etc.).

The basic format of a bottle filename is `$name-$version.$bottle_tag.$bottle_extension`, where `$bottle_extension` is what I've described above. `$bottle_tag` is something like `:catalina`, `:big_sur`, `:arm64_big_sur`, `:x86_64_linux`, etc. Optionally this can be `:all` to indicate that it is os/arch-independent.

A bottle also contains something called a "tab", which is JSON-encoded metadata about the bottle. You can probably get away with not generating one of these for statically linked binaries (such as those generated by Go).

You can download a bottle from the release I linked to above to see what they look like.

**Reply by ryancurrah** [reply]
Seems like one could write a packaging library for bottle in Go similar to nfpm.

I'd be down to try that.

Also this doc seems to help me understand how a bottle works as well: https://docs.brew.sh/Bottles

**Reply by carlocab** [reply]
Sure. But, I want to emphasise what I mentioned above: I can't make any guarantees about this being a stable API to develop against, and this may change without notice. (Changing it will be quite painful, though, so, personally -- i.e. not in my capacity as a Homebrew maintainer, I'd say the API is probably stable enough.)

**Reply by alebcay** [reply]
The thought of a Golang library to support reading/parsing/creating bottles (from arbitrary filesystem contents) that can be read/used/unpacked by Homebrew interests me. I might play with the idea a bit when I have a chance and see where it goes.

**Reply by ryancurrah** [reply]
@carlocab what if you want to provide a brew formula that only has binaries and no option to build from source. Can the `url` be omitted from the formula?

**Reply by SMillerDev** [reply]
> what if you want to provide a brew formula that only has binaries and no option to build from source. Can the url be omitted from the formula?

That's a cask, which will still need some way to fetch the objects.
